File: src/spade/reporter/audit/kernel-modules/spade/audit/state/syscall/syscall.c
/*
 --------------------------------------------------------------------------------
 SPADE - Support for Provenance Auditing in Distributed Environments.
 Copyright (C) 2025 SRI International

 This program is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as
 published by the Free Software Foundation, either version 3 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see <http://www.gnu.org/licenses/>.
 --------------------------------------------------------------------------------
 */


#include <linux/atomic.h>
#include <linux/errno.h>
#include <linux/kernel.h>
#include <linux/version.h>

#include "spade/audit/state/syscall/syscall.h"
#include "spade/audit/helper/kernel.h"
#include "spade/util/log/log.h"


int state_syscall_is_initialized(
    bool *dst,
    struct state_syscall *s
)
{
    if (!dst || !s)
        return -EINVAL;

    if (!s->initialized)
    {
        *dst = false;
        return 0;
    }

    *dst = true;
    return 0;
}

int state_syscall_init(
    struct state_syscall *s, bool dry_run
)
{
    const char *log_id = "state_syscall_init";
    int err;
    bool namespace_is_inited = false;
    bool syscall_is_inited = false;

    if (!s)
        return -EINVAL;

    if (s->initialized)
        return -EALREADY;

    err = state_syscall_namespace_is_initialized(&namespace_is_inited, &s->ns);
    if (err != 0)
        return err;

    if (!namespace_is_inited)
    {
        util_log_debug(log_id, "Initing syscall namespace state");
        err = state_syscall_namespace_init(&s->ns, dry_run);
        if (err != 0)
        {
            util_log_debug(log_id, "Initing syscall namespace state. Failed. Err: %d", err);
            return err;
        } else
        {
            util_log_debug(log_id, "Initing syscall namespace state. Success");
        }
    }

    err = state_syscall_hook_is_initialized(&syscall_is_inited, &s->hook);
    if (err != 0)
        return err;

    if (!syscall_is_inited)
    {
        util_log_debug(log_id, "Initing syscall hook state");
        err = state_syscall_hook_init(&s->hook, dry_run);
        if (err != 0)
        {
            util_log_debug(log_id, "Initing syscall hook state. Failed. Err: %d", err);
            return err;
        } else
        {
            util_log_debug(log_id, "Initing syscall hook state. Success");
        }
    }

    s->initialized = true;
    s->dry_run = dry_run;

    return 0;
}

int state_syscall_deinit(struct state_syscall *s)
{
    int ns_err;
    int sys_err;

    if (!s || !s->initialized)
        return -EINVAL;

    ns_err = state_syscall_namespace_deinit(&s->ns);
    sys_err = state_syscall_hook_deinit(&s->hook);

    s->initialized = false;
    s->dry_run = false;

    if (ns_err != 0)
        return ns_err;

    if (sys_err != 0)
        return sys_err;

    return 0;
}

File: src/spade/reporter/audit/linux/event/record/type/Syscall.java
/*
 --------------------------------------------------------------------------------
 SPADE - Support for Provenance Auditing in Distributed Environments.
 Copyright (C) 2026 SRI International

 This program is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as
 published by the Free Software Foundation, either version 3 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see <http://www.gnu.org/licenses/>.
 --------------------------------------------------------------------------------
 */
package spade.reporter.audit.linux.event.record.type;

import java.util.Map;

import spade.reporter.audit.linux.event.ID;
import spade.reporter.audit.linux.event.record.MalformedRecordException;
import spade.reporter.audit.linux.event.record.Record;
import spade.reporter.audit.linux.event.record.Type;
import spade.reporter.audit.linux.event.record.helper.AuditStringParser;
import spade.reporter.audit.linux.event.record.helper.Header;
import spade.reporter.audit.linux.event.record.helper.KeyValueParser;
import spade.reporter.audit.linux.event.record.helper.ProcessInfo;

/**
 * Record subclass for SYSCALL audit records.
 *
 * Contains process and syscall fields parsed from the audit data.
 *
 * Example raw data:
 *   arch=c000003e syscall=0 success=yes exit=30 a0=6 a1=7fff06b61700 a2=1000 a3=0 items=0
 *   ppid=26414 pid=26415 auid=1000 uid=1002 gid=1002 euid=1002 suid=1002 fsuid=1002
 *   egid=1002 sgid=1002 fsgid=1002 tty=(none) ses=3 comm="screen" exe="/usr/bin/screen" key=(null)
 */
public class Syscall extends Record{

	/** Syscall number. */
	public final String syscall;
	/** Whether the syscall succeeded. */
	public final String success;
	/** Return value / exit code of the syscall. */
	public final String exit;
	/** Syscall argument 0. */
	public final String arg0;
	/** Syscall argument 1. */
	public final String arg1;
	/** Syscall argument 2. */
	public final String arg2;
	/** Syscall argument 3. */
	public final String arg3;
	/** Number of PATH records associated with this syscall. */
	public final String items;
	/** Path of the executable. */
	public final String exe;
	/** Process identity fields. */
	public final ProcessInfo processInfo;

	public Syscall(
		final ID id, final String rawRecord
	) throws MalformedRecordException{
		super(id, Type.SYSCALL, rawRecord);
		final Map<String, String> parsedFields = KeyValueParser.parseKeyValuePairs(rawRecord);
		this.syscall = parsedFields.get("syscall");
		this.success = parsedFields.get("success");
		this.exit = parsedFields.get("exit");
		this.arg0 = parsedFields.get("a0");
		this.arg1 = parsedFields.get("a1");
		this.arg2 = parsedFields.get("a2");
		this.arg3 = parsedFields.get("a3");
		this.items = parsedFields.get("items");
		this.exe = parsedFields.get("exe");
		this.processInfo = new ProcessInfo(
			parsedFields.get("pid"),
			parsedFields.get("ppid"),
			parsedFields.get("uid"),
			parsedFields.get("euid"),
			parsedFields.get("suid"),
			parsedFields.get("fsuid"),
			parsedFields.get("gid"),
			parsedFields.get("egid"),
			parsedFields.get("sgid"),
			parsedFields.get("fsgid"),
			AuditStringParser.mustParse(rawRecord, "comm")
		);
	}

	public String getSyscall(){
		return syscall;
	}

	public String getSuccess(){
		return success;
	}

	public String getExit(){
		return exit;
	}

	public String getArg0(){
		return arg0;
	}

	public String getArg1(){
		return arg1;
	}

	public String getArg2(){
		return arg2;
	}

	public String getArg3(){
		return arg3;
	}

	public String getPid(){
		return processInfo.pid;
	}

	public String getPpid(){
		return processInfo.ppid;
	}

	public String getUid(){
		return processInfo.uid;
	}

	public String getEuid(){
		return processInfo.euid;
	}

	public String getSuid(){
		return processInfo.suid;
	}

	public String getFsuid(){
		return processInfo.fsuid;
	}

	public String getGid(){
		return processInfo.gid;
	}

	public String getEgid(){
		return processInfo.egid;
	}

	public String getSgid(){
		return processInfo.sgid;
	}

	public String getFsgid(){
		return processInfo.fsgid;
	}

	public String getComm(){
		return processInfo.comm;
	}

	public String getExe(){
		return exe;
	}

	public String getItems(){
		return items;
	}

	public static class Creator extends Record.Creator{

		@Override
		public String validate(final Header header){
			final Type type = header.getType();
			return type == Type.SYSCALL ? null : "Expected SYSCALL, got: " + type;
		}

		@Override
		public Record create(final Header header) throws MalformedRecordException{
			final String error = validate(header);
			if(error != null) throw new MalformedRecordException(error, header.getRawLine());
			return new Syscall(header.getId(), header.getRawLine());
		}
	}
}
